polygeist/test/polygeist-opt/sycl/argpromotion.mlir

```diff
@@ -57,11 +57,8 @@ gpu.module @device_func {
     // CHECK-NEXT:    %c1 = arith.constant 1 : index
     // CHECK-NEXT:    [[ARG1:%.*]] = "polygeist.subindex"(%arg0, %c1) : (memref<?x!llvm.struct<(i32, i64)>>, index) -> memref<?xi64>
     // CHECK-NEXT:    {{.*}} = call @callee1([[ARG0]], [[ARG1]]) : (memref<?xi32>, memref<?xi64>) -> i64
-    %alloca = memref.alloca() : memref<i64>        
     %0 = func.call @callee1(%arg0) : (memref<?x!llvm.struct<(i32, i64)>>) -> i64
-    %1 = memref.load %alloca[] : memref<i64>
```

Comment:
open for discussion, should we allow non-aliased memory operation after the call?
Having non-aliased memory operation after the call should be safe for the transformation. 
However, if we only want to perform the transformation on kernel body function, then we can limit candidate to be only on tail calls.
Personally prefer to simplify and limit argument promotion to only perform on tail calls.

Reply:
Have we encountered any actual SYCL codes where the SYCL kernel function is called from a non-tail position? 

Naively, I would assume the only call to SYCL kernel functions can happen from the wrapper function and that should usually be a tail call.

Reply:
I wrote this explicitly to allow operations after the call that aren't aliased. If there aren't any we are OK. Removing this reduces functionality and if we ever end up having a non-aliased load/store after the tail call this code would be needed.  I prefer to leave it as is.

Reply:
IMHO, it's only worth to keep as is if we can actually observe this with SYCL code. Otherwise, keeping the code adds complexity and additional compilation time overhead without any benefit.

Reply:
added back code in argument promotion to allow memory operations after the call.